### PR TITLE
Step 2 API Layer

### PR DESCRIPTION
**Pull request by a-reznikov**
image
image

- [x] Look at NYCTimes API -[ https://developer.nytimes.com/docs/top-stories-product/1/overview](https://developer.nytimes.com/docs/top-stories-product/1/overview)
- [x] Create account and app on site to generate your private API key
- [x] Try your key with top-stories-product API
- [x] Read about https://tanstack.com/query/latest
- [x] Install react query to your application
- [x] Choose section that you like from NYC Top Stories API and make request from your app
- [x] Create component to list stories from API
- [x] Create commit and PR to main branch

### CODE REVIEW

File: public/menu.svg
Comment (Ilya-Valasiuk) [inline public/menu.svg:1]: public/icons/

File: public/news.jpg
Comment (Ilya-Valasiuk) [inline public/news.jpg:1]: public/backgrounds/

File: src/pages/_app.tsx [hunk 1]
```diff
@@ -1,11 +1,16 @@
+import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
 import type { AppProps } from "next/app";
 import { OPEN_SANS } from "@/fonts";
 import "@/styles/globals.css";
 
+const queryClient = new QueryClient();
```
Comment (Ilya-Valasiuk) [inline src/pages/_app.tsx]: move to src/constants/query-client.ts

File: src/components/home/Home.tsx [hunk 2]
```diff
@@ -1,7 +1,9 @@
+import { NewsList } from "../common/news-list/NewsList";
+
 export const Home: React.FC = () => {
   return (
     <div className="w-full py-10">
-      <h1>Main Section</h1>
+      <NewsList />
```
Comment (Ilya-Valasiuk) [inline src/components/home/Home.tsx]: ```<NewsList section="home" />```

File: src/components/common/news-list/NewsList.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { useQuery } from "@tanstack/react-query";
+import { NewsCard } from "../news-card/NewsCard";
+import { getNewsBySection } from "@/api/news-section/module";
+import { HEADER_NAVIGATION } from "@/constants/links";
+
+export const NewsList: React.FC = () => {
+  const sectionName = HEADER_NAVIGATION[0].href.slice(1);
```
Comment (Ilya-Valasiuk) [inline src/components/common/news-list/NewsList.tsx]: dangerous

File: src/components/common/news-card/NewsCard.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import Image from "next/image";
+import classNames from "classnames";
+import { News } from "../../../api/news-section/types";
+
```
Comment (Ilya-Valasiuk) [inline src/components/common/news-card/NewsCard.tsx:4]:
```
type Props = {
  news: News
}

export const NewsCard: React.FC<Props>
```

File: src/components/common/news-card/NewsCard.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        </p>
+        <div className="flex flex-col gap-4">
+          <h2 className="text-xl font-semibold">{title}</h2>
+          <p>{abstract}</p>
+        </div>
+      </div>
+      <Image
+        className={classNames(
+          "w-full basis-full object-cover min-h-70 max-h-70",
+          "lg:w-0 lg:basis-51 lg:shrink-0 lg:min-h-114.5 lg:max-h-114.5"
```
Comment (Ilya-Valasiuk) [inline src/components/common/news-card/NewsCard.tsx]:
```
"lg:w-0 lg:basis-[51%] lg:shrink-0 lg:min-h-[458px] lg:max-h-[280px]"
```

File: src/components/common/header/Header.tsx [hunk 6]
```diff
@@ -33,6 +36,9 @@ export const Header: React.FC = () => {
             </a>
           ))}
         </nav>
+        <a href="/" className={classNames("my-4", "sm:hidden")}>
+          <Image src="/menu.svg" width={28} height={28} alt="Menu" priority />
```
Comment (Ilya-Valasiuk) [inline src/components/common/header/Header.tsx]: priority?

Comment (Ilya-Valasiuk) [inline src/components/common/header/Header.tsx]: https://react-svgr.com/docs/next/

File: src/api/news-section/types.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+export interface NewsImage {
+  url: string;
+  type: string;
+}
+
+export interface RawNews {
```
Comment (Ilya-Valasiuk) [inline src/api/news-section/types.ts]: type

Comment (Ilya-Valasiuk) [reply src/api/news-section/types.ts]:
```
NewsRaw
```

File: src/api/news-section/types.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  section: string;
+  updated_date: string;
+  title: string;
+  abstract: string;
+  uri: string;
+  multimedia: NewsImage[];
+}
+
+export interface Data {
```
Comment (Ilya-Valasiuk) [inline src/api/news-section/types.ts]: export type NycTimesResponse

File: src/api/news-section/types.ts (hunk 7 above)
Comment (Ilya-Valasiuk) [reply src/api/news-section/types.ts]: News

File: src/api/news-section/module.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        abstract,
+        uri,
+        imgUrl: multimedia.length ? multimedia[0].url : "/news.jpg",
+        imgAlt: multimedia.length ? multimedia[0].type : "news image",
+      };
+    }
+  );
+};
+
+export const getNewsBySection = async (
```
Comment (Ilya-Valasiuk) [inline src/api/news-section/module.ts]:
```
src/api/news
  - module.ts (only fetch)
  - queris.ts (ony react queries)
  - types.ts
  - formatter.ts
```

quires.ts

```
import { useQuery } from '@tanstack/react-query';
import { fetchCompanyOptions } from './module';

export enum QueryKeys {
  USER_COMPANIES = 'use-companies',
}

export const useUserCompanyOptionsToSelect = (enabled = false) =>
  useQuery({
    queryKey: [QueryKeys.USER_COMPANIES],
    queryFn: fetchCompanyOptions,
    staleTime: Infinity,
    enabled,
  });

```

File: public/menu.svg
Comment (a-reznikov) [reply public/menu.svg:1]: Added

File: public/news.jpg
Comment (a-reznikov) [reply public/news.jpg:1]: Added

File: src/pages/_app.tsx (hunk 1 above)
Comment (a-reznikov) [reply src/pages/_app.tsx]: Created

File: src/components/home/Home.tsx (hunk 2 above)
Comment (a-reznikov) [reply src/components/home/Home.tsx]: Added

File: src/components/common/news-list/NewsList.tsx (hunk 3 above)
Comment (a-reznikov) [reply src/components/common/news-list/NewsList.tsx]: Fixed

File: src/components/common/news-card/NewsCard.tsx (hunk 4 above)
Comment (a-reznikov) [reply src/components/common/news-card/NewsCard.tsx:4]: Fixed

File: src/api/news-section/types.ts (hunk 7 above)
Comment (a-reznikov) [reply src/api/news-section/types.ts]: Fixed

File: src/api/news-section/types.ts (hunk 8 above)
Comment (a-reznikov) [reply src/api/news-section/types.ts]: Renamed

File: src/api/news-section/module.ts (hunk 9 above)
Comment (a-reznikov) [reply src/api/news-section/module.ts]: Implemented

File: src/components/common/news-card/NewsCard.tsx (hunk 5 above)
Comment (a-reznikov) [reply src/components/common/news-card/NewsCard.tsx]: Fixed

File: src/components/common/header/Header.tsx (hunk 6 above)
Comment (a-reznikov) [reply src/components/common/header/Header.tsx]: Removed

Comment (a-reznikov) [reply src/components/common/header/Header.tsx]: Used

File: src/api/news/module.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { NycTimesResponse, News } from "@/api/news/types";
+import { prepareData } from "./formatter";
+
+export const getNewsBySection = async (
+  sectionName: string
+): Promise<News[]> => {
+  const response: Response = await fetch(
+    `https://api.nytimes.com/svc/topstories/v2/${sectionName}.json?api-key=${process.env.NEXT_PUBLIC_API_KEY}`
```
Comment (Ilya-Valasiuk) [inline src/api/news/module.ts]: https://api.nytimes.com -> env variables

src/constants/variables.ts

export const API_KEY = process.env.NEXT_PUBLIC_API_KEY;

if (!API_KEY) {
  console.error()
} 

export const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL;

File: src/constants/links.ts [hunk 11]
```diff
@@ -1,8 +1,9 @@
 export const HEADER_NAVIGATION = [
```
Comment (Ilya-Valasiuk) [inline src/constants/links.ts:5]: import X from '/icons/twitter.svg';


.jsx
```
<X className="text-white" />
```

jsx
```
{React.createElement(X, { className: 'text-white'})}
```

```
export const SOCIAL_LINKS = [
  { alt: "Twitter", component: X, href: "#" },
];
```

```
{SOCIAL_LINKS.map({ component } => <div>React.createElement(component, { className: 'text-white' })</div>)}
```

File: src/api/news/module.ts (hunk 10 above)
Comment (a-reznikov) [reply src/api/news/module.ts]: Implemented

File: src/constants/links.ts (hunk 11 above)
Comment (a-reznikov) [reply src/constants/links.ts:5]: Implemented

**Review by Ilya-Valasiuk — APPROVED** [review]

**Pull request merged**